File: frontend/types/trip.ts
/**
 * Trip domain types.
 *
 * These mirror the Sprint 1 backend trip CRUD API (FastAPI + Pydantic).
 * Read shapes match `TripRead` / `TripPreferenceRead`; input shapes match
 * `TripCreate` / `TripUpdate`.
 *
 * Notes on field formats coming from the backend:
 * - `start_date` / `end_date` are ISO date strings (`YYYY-MM-DD`) or null.
 * - `created_at` / `updated_at` are ISO datetime strings.
 * - `budget` is a Pydantic `Decimal`, which serializes to a JSON *string*
 *   (e.g. "1500.00") on read. On input a number is accepted and coerced.
 */

/** Trip lifecycle. Matches backend `TripStatus`. */
export type TripStatus = "draft" | "planned" | "completed" | "archived";

/** Shared trip preference fields (one-to-one with a trip). */
export interface TripPreferenceBase {
  travel_style: string | null;
  max_walking_minutes_between_stops: number | null;
  max_total_walking_minutes_per_day: number | null;
  interests: string[];
  hotel_preferences: string[];
  food_preferences: string[];
  must_visit_places: string[];
  avoid_places: string[];
}

/** Trip preference as returned by the API (`TripPreferenceRead`). */
export interface TripPreference extends TripPreferenceBase {
  id: number;
  trip_id: number;
}

/** A trip as returned by the API (`TripRead`). */
export interface Trip {
  id: number;
  user_id: number;
  title: string;
  destination: string;
  start_date: string | null;
  end_date: string | null;
  travelers: number;
  budget: string | null;
  status: TripStatus;
  created_at: string;
  updated_at: string;
  preference: TripPreference | null;
}

/** Preference fields when creating/updating a trip (`TripPreferenceCreate`). */
export interface TripPreferenceInput {
  travel_style?: string | null;
  max_walking_minutes_between_stops?: number | null;
  max_total_walking_minutes_per_day?: number | null;
  interests?: string[];
  hotel_preferences?: string[];
  food_preferences?: string[];
  must_visit_places?: string[];
  avoid_places?: string[];
}

/** Request body for creating a trip (`TripCreate`). */
export interface CreateTripInput {
  title: string;
  destination: string;
  start_date?: string | null;
  end_date?: string | null;
  travelers?: number;
  budget?: number | string | null;
  status?: TripStatus;
  preference?: TripPreferenceInput | null;
}

/** Request body for a partial update (`TripUpdate`). All fields optional. */
export type UpdateTripInput = Partial<CreateTripInput>;

File: backend/app/schemas/trip.py
"""Trip and TripPreference API schemas.

Naming convention:
- ``*Base``   shared fields
- ``*Create`` request body for creation (no server-set fields)
- ``*Update`` request body for partial update (all fields optional)
- ``*Read``   response body (includes server-set fields like ids/timestamps)
"""

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.trip import TripStatus


# --- TripPreference ---------------------------------------------------------


class TripPreferenceBase(BaseModel):
    travel_style: str | None = None
    max_walking_minutes_between_stops: int | None = Field(default=None, ge=0)
    max_total_walking_minutes_per_day: int | None = Field(default=None, ge=0)
    interests: list[str] = Field(default_factory=list)
    hotel_preferences: list[str] = Field(default_factory=list)
    food_preferences: list[str] = Field(default_factory=list)
    must_visit_places: list[str] = Field(default_factory=list)
    avoid_places: list[str] = Field(default_factory=list)


class TripPreferenceCreate(TripPreferenceBase):
    pass


class TripPreferenceRead(TripPreferenceBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    trip_id: int


# --- Trip -------------------------------------------------------------------


class TripBase(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    destination: str = Field(min_length=1, max_length=200)
    start_date: date | None = None
    end_date: date | None = None
    travelers: int = Field(default=1, ge=1)
    budget: Decimal | None = Field(default=None, ge=0)
    status: TripStatus = TripStatus.DRAFT


class TripCreate(TripBase):
    # Preferences may be supplied inline when creating a trip.
    preference: TripPreferenceCreate | None = None


class TripUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    destination: str | None = Field(default=None, min_length=1, max_length=200)
    start_date: date | None = None
    end_date: date | None = None
    travelers: int | None = Field(default=None, ge=1)
    budget: Decimal | None = Field(default=None, ge=0)
    status: TripStatus | None = None
    preference: TripPreferenceCreate | None = None


class TripRead(TripBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime
    preference: TripPreferenceRead | None = None
